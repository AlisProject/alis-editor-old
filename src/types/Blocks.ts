export enum BlockType {
  Rule = 'Rule',
  Text = 'Text',
  Paragraph = 'Paragraph',
  Image = 'Image',
  Quote = 'Quote',
  Heading = 'Heading',
  Embed = 'Embed'
}

export interface Block {
  id: string
  type: BlockType
  payload?: any
  children?: Block[]
}

export interface RuleBlock extends Block {
  type: BlockType.Rule
}

export interface EmbedBlock extends Block {
  type: BlockType.Embed,
  payload: {
    src: string
  }
}

export interface ParagraphBlock extends Block {
  type: BlockType.Paragraph
  payload: {
    body: string
  }
}

export interface TextBlock extends Block {
  type: BlockType.Text
  payload: {
    body: string
  }
}

export interface ImageBlock extends Block {
  type: BlockType.Image
  payload: {
    src: string
    align?: string
  }
  children: Block[]
}

export interface HeadingBlock extends Block {
  type: BlockType.Heading
  payload: {
    size: 'h2' | 'h3'
    body: string
  }
  children: Block[]
}

export interface QuoteBlock extends Block {
  type: BlockType.Quote
  children: Block[]
}

/*
「詳細は [こちら](https://example.com) から確認できます」を再現する場合
[
  {
    type: 'Paragraph',
    children: [
      {
        type: 'Text',
        payload: {
          body: '詳細は'
        }
      },
      {
        type: 'Link',
        payload: {
          href: 'https://example.com'
        },
        children: [
          {
            type: 'Text',
            payload: {
              body: 'こちら'
            }
          }
        ]
      },
      {
        type: 'Text',
        payload: {
          body: 'から確認できます'
        }
      }
    ]
  }
]

*/
