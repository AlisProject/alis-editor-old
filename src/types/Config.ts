export interface AlisEditorConfig {
  articleId: string
  preview: boolean
  iframelyApikey: string
  uploadEndpoint: string
}
