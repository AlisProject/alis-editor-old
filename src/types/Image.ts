export interface ImageUploadRequest {
  article_image: string
}

export interface ImageUploadResponse {
  image_url: string
}
