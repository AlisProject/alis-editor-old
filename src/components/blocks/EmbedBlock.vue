<template>
  <div>
    <div v-html="html"></div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { EmbedBlock } from '../../types/Blocks'
import ShadowInput from '../utils/ShadowInput.vue'
import {
  isYouTubeVideoURL,
  isFacebookPostURL,
  isInstagramURL,
  getResourceFromIframely,
  getIframelyUrlTemplate,
  getTwitterProfileTemplate,
  getIframelyEmbedTemplate
} from '../../utils/iframely'

export default Vue.extend({
  components: {
    ShadowInput
  },
  props: {
    block: Object as () => EmbedBlock,
    iframelyApikey: String
  },
  data() {
    return {
      isFocus: false,
      html: ''
    }
  },
  async mounted() {
    this.html = await this.getEmbedContent(this.block.payload.src)
    ;(window as any).iframely.load()
  },
  methods: {
    handleFocus() {
      this.isFocus = true
    },
    handleBlur() {
      this.isFocus = false
    },
    handleDelete() {
      this.$emit('delete', this.block)
    },
    async getEmbedContent(alisIframelyUrl: string) {
      const isTwitterResource =
        alisIframelyUrl === 'https://twitter.com' ||
        alisIframelyUrl.startsWith('https://twitter.com/')
      const isTweet = isTwitterResource && alisIframelyUrl.split('/')[4] === 'status'
      const isGistResource = alisIframelyUrl.startsWith('https://gist.github.com/')
      const isYouTubeResource = isYouTubeVideoURL(alisIframelyUrl)
      const isFacebookResource = isFacebookPostURL(alisIframelyUrl)
      const isInstagramResource = isInstagramURL(alisIframelyUrl)
      let result

      try {
        result = (await getResourceFromIframely(
          {
            type: isTwitterResource ? 'oembed' : 'iframely',
            url: alisIframelyUrl,
            iframelyApikey: this.iframelyApikey
          }
        )).data
      } catch (error) {
        console.error(error)
        return ''
      }

      if (
        isTweet ||
        isGistResource ||
        isYouTubeResource ||
        isFacebookResource ||
        isInstagramResource
      ) {
        const anchorElement = document.createElement('a')
        anchorElement.setAttribute('href', alisIframelyUrl)
        anchorElement.setAttribute('data-iframely-url', '')

        const div = document.createElement('div')
        div.appendChild(anchorElement)

        return div.innerHTML
      }

      if (isTwitterResource) {
        const { title, description } = result
        const hasTitleOrDescription = title !== undefined || description !== undefined
        if (!hasTitleOrDescription) return ''

        return getTwitterProfileTemplate({ ...result })
      } else {
        const { title, description } = result.meta
        const hasTitleOrDescription = title !== undefined || description !== undefined
        if (!hasTitleOrDescription) return ''

        return getIframelyEmbedTemplate({ ...result })
      }
    }
  }
})
</script>

<style scoped>
</style>
