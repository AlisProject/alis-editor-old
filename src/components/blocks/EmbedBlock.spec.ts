import { mount, config } from '@vue/test-utils'
import EmbedBlock from './EmbedBlock.vue'
import * as BlockTypes from '../../types/Blocks'
config.logModifiedComponents = false

describe('EmbedBlock.vue', () => {
  test('renders correctly', () => {
    const block: BlockTypes.EmbedBlock = {
      id: '123',
      type: BlockTypes.BlockType.Embed,
      payload: {
        src: 'https://twitter.com/_/status/1'
      },
      children: []
    }
    const wrapper = mount(EmbedBlock, {
      propsData: {
        block
      },
      stubs: ['ShadowInput']
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders twitter profile content correctly', () => {
    const block: BlockTypes.EmbedBlock = {
      id: '123',
      type: BlockTypes.BlockType.Embed,
      payload: {
        src: 'https://twitter.com/ALIS_media'
      },
      children: []
    }
    const wrapper = mount(EmbedBlock, {
      propsData: {
        block
      },
      stubs: ['ShadowInput']
    })
    wrapper.setData({
      html: `<div data-alis-iframely-url="https://twitter.com/alis_media" contenteditable="false">
        <a href="https://twitter.com/alis_media" target="_blank" class="twitter-profile-card">
          <div class="title">ALIS (@ALIS_media) | Twitter</div><div class="description">
            The latest Tweets from ALIS (@ALIS_media). ALIS is social media platform using blockchain. We help people to find trustworthy articles and people quickly.
            ALIS公式アカウントです。日々の活動状況やニュースを発信します。#ALIS $ALIS. Japan
          </div>
          <div class="site">twitter.com</div>
        </a>
      </div>`
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders OGP content correctly', () => {
    const block: BlockTypes.EmbedBlock = {
      id: '123',
      type: BlockTypes.BlockType.Embed,
      payload: {
        src: 'https://alis.to'
      },
      children: []
    }
    const wrapper = mount(EmbedBlock, {
      propsData: {
        block
      },
      stubs: ['ShadowInput']
    })
    wrapper.setData({
      html: `<div data-alis-iframely-url="https://alis.to" contenteditable="false">
        <a href="https://alis.to" target="_blank" class="iframely-embed-card">
          <div class="title">ALIS | 信頼できる記事と人々を明らかにする全く新しいソーシャルメディア</div>
          <div class="description">クリプトの人気記事一覧</div>
          <img src="https://alis.to/d/nuxt/dist/touch-icon.png" class="thumbnail">
          <div class="site">alis.to</div>
        </a>
      </div>`
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
