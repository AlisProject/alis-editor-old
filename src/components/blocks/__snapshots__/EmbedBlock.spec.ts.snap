// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmbedBlock.vue renders correctly 1`] = `
<div>
  <div>
    <br />
    <div
      contenteditable="false"
      data-alis-iframely-url="https://twitter.com/_/status/1"
    >
      <a
        data-iframely-url=""
        href="https://twitter.com/_/status/1"
      />
    </div>
    <br />
  </div>
</div>
`;
